### DISCUSSION TITLE
Problem in accessing javascript scope from python

### DISCUSSION THREAD

**Discussion by bajrangCoder** [pyodide/pyodide#3998] [Q&A] [answered]
I am creating formatter using black and i have used pyodide to run python on browser.
But there is issue in accessing javascript scope inside python .
I have loaded all required things here:
```js
async function loadPyodideAndPackages(baseUrl) {
    importScripts(`${baseUrl}/lib/pyodide.js`);
    self.pyodide = await loadPyodide();
    await self.pyodide.loadPackage(`${baseUrl}/lib/packaging-23.0-py3-none-any.whl`);
    await self.pyodide.loadPackage(`${baseUrl}/lib/black-23.7.0-py3-none-any.whl`);
    await self.pyodide.loadPackage(`${baseUrl}/lib/platformdirs-3.8.1-py3-none-any.whl`);
    await self.pyodide.loadPackage(`${baseUrl}/lib/mypy_extensions-1.0.0-py3-none-any.whl`);
    await self.pyodide.loadPackage(`${baseUrl}/lib/pathspec-0.11.1-py3-none-any.whl`);
    await self.pyodide.loadPackage(`${baseUrl}/lib/click-8.1.3-py3-none-any.whl`);
    console.log("Loaded all required modules 🎉");
}
```
Here `baseUrl` is the base location where files are stored. loading of files is working I also get console log message .

And here I am trying to get javascript scope:

> sourceCode variable value inside the python scope

```js
async formatCode(data) {
        const { sourceCode } = data;
        try {
            await self.pyodide.runPythonAsync(`
                import black
                from js import sourceCode
                result = {}
                print(sourceCode)
                if sourceCode.value == "":
                    result["message"] = "No source code provided"
                else:
                    try:
                        mode = black.FileMode()
                        result["formatted"] = black.format_file_contents(sourceCode.value, fast=False, mode=mode)
                    except black.NothingChanged:
                        result["formatted"] = sourceCode.value
                    except Exception as exc:
                        result["message"] = str(exc)
            `);
            let res = await self.pyodide.globals.get("result").toJs();
            self.postMessage({
                action: 'formatCode',
                success: true,
                output: res,
            });
        } catch(error) {
            self.postMessage({
                action: 'formatCode',
                success: false,
                error: error?.message ?? error?.toString(),
            });
        }
    }
```

But it gives error following error:

![Screenshot_20230716-092344_1](https://github.com/pyodide/pyodide/assets/71929976/88968709-6064-4d14-80f4-e8cbe06533a0)

> Note: I am using web workers

What's the problem? Please help me

**Comment by hoodmane** [comment, marked as answer]
`js` is `globalThis`. If you want to use `from js import sourceCode` say `globalThis.sourceCode = sourceCode`. There are a few other ways not using `from js import x`:
1. Set a Python global from a non-global JS variable: `pyodide.globals.set("sourceCode", sourceCode)`
2. Define a function and avoid the use of globals entirely:
```js
pyodide.runPython(`
def doStuff(sourceCode):
    # do stuff with sourceCode
`);
const doStuff = pyodide.globals.get("doStuff");
doStuff(sourceCode)
```

Avoiding the global in both Python and JS is arguable the best coding style.

**Reply by hoodmane** [reply]
I'm closing this as answered but don't hesitate to ask further questions.

**Reply by bajrangCoder** [reply]
Thanks for quick response!
Your answer solved my issued
❤️❤️

**Comment by bajrangCoder** [comment]
One more question: Is pyodide have any problem on running on old Android version like 9 etc.
Because when i run pyodide with given cdn then it works fine but when i use all pyodide files from local and run it through web workers then it fails with this error 👇
![IMG_20230716_134922_563~2](https://github.com/pyodide/pyodide/assets/71929976/fbf127d5-c918-4d93-9e49-469b2c164da5)

**Reply by hoodmane** [reply]
Go to chrome://version/, what does it say there? You might also try running in Firefox in case you can get a newer Firefox than chrome. Currently I think we require chrome 85, Firefox 78, Safari 14.1. I think those have been the min versions for quite a while though I'd like to step the min requirements to chrome 96, Firefox 100, Safari 15.2 soon.

**Reply by bajrangCoder** [reply]
Is there any minimum version for Android web view because it's problem in only web view, it's works on browsers.

**Reply by hoodmane** [reply]
I don't know anything about Android web view. Can you open an issue with a reproduction? Do you know what version of v8/chrome it uses? Perhaps there's a problem with the environment detection code.

**Reply by bajrangCoder** [reply]
For reproduction just load pyodide from local files and instead of running it on browser run it with web view. Note: It's works fine on Android 10 but not on Android 9
